### PR TITLE
Adds support for Rulesets having intrinsic template requirements

### PR DESCRIPTION
**Pull request by ZephyrTFA**
Title

## About The Pull Request

Ensures that we load templates for a ruleset before we attempt to place or cache characters for that ruleset
Also makes wizard and abductor async load their template to improve (apparent) loading times for them
## Why It's Good For The Game

This is the only thing left that I can think of that would cause antags like nukies and abductors to spawn in wrong
## Changelog
This should not be player facing

### CODE REVIEW

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 1]
```diff
@@ -198,11 +199,13 @@
...
 
-/datum/dynamic_ruleset/midround/from_living/ready(forced)
-	if(!check_candidates())
-		return FALSE
-	return ..()
+/datum/dynamic_ruleset/midround/from_living/check_candidates()
+	if (required_candidates <= living_players.len)
+		return TRUE
+
+	log_dynamic("FAIL: [src], a from_ghosts ruleset, did not have enough dead candidates: [required_candidates] needed, [dead_count] found")
```
Comment (ZephyrTFA) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:206]:
```suggestion
	log_dynamic("FAIL: [src], a from_living ruleset, did not have enough candidates: [required_candidates] needed, [living_players.len] found")
```

File: code/__DEFINES/dynamic.dm [hunk 2]
```diff
@@ -7,6 +7,9 @@
 /// This ruleset can only be picked once. Anything that does not have a scaling_cost MUST have this.
 #define LONE_RULESET (1 << 2)
 
+/// This ruleset has lazy loaded templates that must be loaded if the ruleset is selected
+#define RULESET_USES_LAZY_TEMPLATES (1 << 3)
```
Comment (Mothblocks) [inline code/__DEFINES/dynamic.dm:11]: This flag doesn't feel necessary

File: code/game/gamemodes/dynamic/dynamic_rulesets.dm [hunk 3]
```diff
@@ -170,8 +173,12 @@
 /// Here you can perform any additional checks you want. (such as checking the map etc)
 /// Remember that on roundstart no one knows what their job is at this point.
 /// IMPORTANT: If ready() returns TRUE, that means pre_execute() or execute() should never fail!
-/datum/dynamic_ruleset/proc/ready(forced = 0)
-	return check_candidates()
```
Comment (Mothblocks) [inline code/game/gamemodes/dynamic/dynamic_rulesets.dm:174]: Hmmmm are we sure that forcing everything to call this doesn't change behavior at all? I am fairly sure some of the calls without were intentional

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/__DEFINES/dynamic.dm (hunk 2 above)
Comment (ZephyrTFA) [reply code/__DEFINES/dynamic.dm:11]: I can have it just check for the list existing?

File: code/game/gamemodes/dynamic/dynamic_rulesets.dm (hunk 3 above)
Comment (ZephyrTFA) [reply code/game/gamemodes/dynamic/dynamic_rulesets.dm:174]: how about just loading the templates before it calls ready?

File: code/__DEFINES/dynamic.dm (hunk 2 above)
Comment (Mothblocks) [reply code/__DEFINES/dynamic.dm:11]: You don't even haev to do that, looping over null does nothing

File: code/game/gamemodes/dynamic/dynamic_rulesets.dm (hunk 3 above)
Comment (Mothblocks) [reply code/game/gamemodes/dynamic/dynamic_rulesets.dm:174]: Yeah

File: code/controllers/subsystem/mapping.dm [hunk 4]
```diff
@@ -853,6 +853,16 @@ GLOBAL_LIST_EMPTY(the_station_areas)
 
 /datum/controller/subsystem/mapping/proc/lazy_load_template(template_key, force = FALSE)
 	RETURN_TYPE(/datum/turf_reservation)
+	var/static/lazy_loading = FALSE
+	UNTIL(!lazy_loading)
+
+	lazy_loading = TRUE
+	. = _lazy_load_template(template_key, force)
+	lazy_loading = FALSE
```
Comment (Mothblocks) [inline code/controllers/subsystem/mapping.dm:861]:
```suggestion
	lazy_loading = FALSE
	return .
```

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm [hunk 5]
```diff
@@ -273,10 +273,12 @@
 	weight = 2
 	cost = 20
 	requirements = list(90,90,90,80,60,40,30,20,10,10)
+	ruleset_lazy_templates = list(LAZY_TEMPLATE_KEY_WIZARDDEN)
 
-/datum/dynamic_ruleset/roundstart/wizard/acceptable(population=0, threat=0)
-	SSmapping.lazy_load_template(LAZY_TEMPLATE_KEY_WIZARDDEN)
-	if(GLOB.wizardstart.len == 0)
+/datum/dynamic_ruleset/roundstart/wizard/ready(forced = FALSE)
+	if(!check_candidates())
```
Comment (Mothblocks) [inline code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:279]: I'm still not sure we should be changing how the rulesets act in this PR, old Dynamic code is pretty clunky and I wouldn't be surprised if this broke something

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Looks good other than this

File: code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm (hunk 5 above)
Comment (ZephyrTFA) [reply code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:279]: Just made it copy the other ruleset logic

**Pull request merged**